proposals/lip-introduce_a_genric_keystore.md

```diff
@@ -0,0 +1,276 @@
+
+In the same encrypted file, we can store metadata indicating how the secret recovery phrase was used and which private keys were already generated with it. This allows users to not only recover all their accounts when importing the encrypted file in a new device, but also to make sure that newly generated private keys are using a new derivation path.
+
+### Encrypting Private Keys
+
+The keystore presented below is also designed to encrypt private keys. The reason to encrypt and store private keys directly is two fold. First it improves the efficiency of the signing process. Indeed, if we decrypt the private key directly, there is no need to derive the key again from the secret recovery phrase. Secondly, in the case the device of the user was corrupted, decrypting just one private key would compromise the account linked to this private key, but not the others generated with the same secret recovery phrase.
+
+
+## Specification
+
+The specifications below are inspired from [Web3 Secret Storage](https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition) with the addition of a `metadata` property which allows to store all needed information regarding the encrypted material.
+
+### Encryption File Format
+
+We store secret recovery phrases and private keys in a JSON file following the format below.
+
+```java
+keystoreSchema = {
+  "type": "object",
+  "required": ["encryptedPassphrase", "metadata", "ID"],
+  "properties": {
+    "encryptedPassphrase": {
+      "type": "object",
+      "properties": {
+        "version": {"type": "string"},
+        "ciphertext": {"type": "string"},
+        "mac": {"type": "string"},
+        "kdf": {"type": "string"},
+        "kdfparams": {"type": "object"},
+        "cipher": {"type": "string"},
+        "cipherparams": {"type": "object"}
+      }
+    },
+    "metadata": {
+        "name": {"type": "string"},
+        "description": {"type": "string"},
+        "pubkey": {"type": "string"},
+        "address": {"type": "string"},
+        "path": {"type": "string"},
+        "derivedFromID": {"type": "string"},
+        "creationTime": {"type": "string"},
+        "pathsUsed": {"type": "array"},
+        "tags": {"type": "array"}
+    },
+    "ID": {
+        "type": "string",
+        "format": "uuid"
+    }
+  }
+}
+```
+
+In the following sections, we describe the uses of the properties of `keystoreSchema`.
+
+#### encryptedPassphrase
+
+##### version
+
+The `version` is set to `"1"`.
+
+##### ciphertext
+
+The encrypted message in hexadecimal format.
+
+##### mac
+
+Computed as `SHA256(encryptionKey[-16:] + cipherText)`, where `encryptionKey` is the key derived from the user password using the key-derivation function. It can be used to verify the encryption key before starting the decryption process.
```

Comment:
user password is first mentioned here. Perhaps add a one-sentence explanation (either here or in the rationale) on what the password to make clear that it's different than the secret recovery phrase.

Reply:
added